packages/dds/merge-tree/src/partialLengths.ts

```diff
@@ -14,35 +14,106 @@ import {
     ISegment,
     toRemovalInfo,
 } from "./mergeTreeNodes";
+import { SortedSet } from "./sortedSet";
 
-interface IOverlapClient {
-    clientId: number;
-    seglen: number;
-}
+class PartialSequenceLengthsSet extends SortedSet<PartialSequenceLength, number> {
+    protected getKey(item: PartialSequenceLength): number {
+        return item.seq;
+    }
 
-/**
- * Returns the partial length whose sequence number is
- * the greatest sequence number within a that is
- * less than or equal to key.
- * @param a - array of partial segment lengths
- * @param key - sequence number
- */
-function latestLEQ(a: PartialSequenceLength[], key: number) {
-    let best = -1;
-    let lo = 0;
-    let hi = a.length - 1;
-    while (lo <= hi) {
-        const mid = lo + Math.floor((hi - lo) / 2);
-        if (a[mid].seq <= key) {
-            if ((best < 0) || (a[best].seq < a[mid].seq)) {
-                best = mid;
+    public addOrUpdate(
+        newItem: PartialSequenceLength,
+        update?: (existingItem: PartialSequenceLength, newItem: PartialSequenceLength) => void,
+    ) {
+        const prev = this.latestLEQ(newItem.seq);
+
+        if (prev?.seq !== newItem.seq) {
+            // new element, update len
+            newItem.len = (prev?.len ?? 0) + newItem.seglen;
+        }
+
+        // update the len of all following elements
+        for (const e of this.keySortedItems) {
```

Comment:
some of the test failures look non-flaky: they are on tests that this code would affect. The choice to update lengths after each addition makes things scale potentially O(collab window^2) which is a problem.

I think we should probably special-case adding to the end of the list in this helper as well, since we build these lists up in a merge-sort fashion.

Reply:
Yeah the newly failing tests are a result of rebasing onto `next`. There was a change to make equality based on object pointer, rather than `key`, which breaks the assumptions made in this PR. It is trivial to bring back the old behavior specifically for the implementation in partial lengths, but I want to see if it's possible to use the old behavior everywhere, as it is more intuitive and would reduce code size.

Reply:
> special-case adding to the end of the list

In 7d431f5, I've reversed the order of iteration, which has the same effect as special casing appends, and also means we will no longer superfluously visit elements.

Reply:
yeah, this is better. We still incur a binary search that we weren't doing before. Not sure how much that matters in practice (tbh I haven't looked into where the hot paths are for shared string, perf-wise, and the scaling here is with collab window which in practice shouldn't get particularly big)